projects/bms_carrier/src/cell_sense.c

```diff
@@ -0,0 +1,87 @@
+#include "cell_sense.h"
+
+#include <string.h>
+
+#include "bms.h"
+#include "bms_events.h"
+#include "critical_section.h"
+#include "current_sense.h"
+#include "exported_enums.h"
+#include "ltc_afe.h"
+#include "status.h"
+#include "thermistor.h"
+
+static CellSenseStorage s_storage = { 0 };
+
+static void prv_extract_cell_result(uint16_t *result_arr, size_t len, void *context) {
+  ltc_afe_request_aux_conversion(s_storage.afe);
+
+  bool disabled = critical_section_start();
+  memcpy(s_storage.readings->voltages, result_arr, sizeof(s_storage.readings->voltages));
+  critical_section_end(disabled);
+
+  bool fault = false;
+  for (size_t i = 0; i < len; i++) {
+    // s_storage.total_voltage += s_storage.readings->voltages[i];
+    if (s_storage.readings->voltages[i] < s_storage.settings.undervoltage_dmv ||
+        s_storage.readings->voltages[i] > s_storage.settings.overvoltage_dmv) {
+      fault = true;
+    }
+  }
+
+  fault_bps(EE_BPS_STATE_FAULT_CURRENT_SENSE_AFE_CELL, !fault);
+}
+
+static void prv_extract_aux_result(uint16_t *result_arr, size_t len, void *context) {
+  ltc_afe_request_cell_conversion(s_storage.afe);
```

Comment:
Do you think constantly switching between conversions would eat up too much compute? would BMS have enough time to work on other things?

Honestly no idea if that's a reasonable concern or not, there's just a lot of other things going on, mainly current sense reads a lot. Otherwise in my head there'd be a small soft timer between voltage/temp readings since they're unlikely to change a lot that rapidly

Reply:
There are delays between various parts of conversion so it should leave a decent amount of time to do other operations. We can add extra delays if these results aren't critical. I was thinking the same thing but it wasn't in the last car so I opted to use the same model.